File: ocaml/src/vectorize/intrinsic_builder.ml
open Llvm

let create intrinsic_builder name operands imm8 =
  let wrapper_name = Printf.sprintf "intrinsic_wrapper_%s_%d" name imm8 in
  let f = intrinsic_builder.inst_wrappers
          |> Function.lookup wrapper_name
          |> Option.get in
  assert (List.length (Function.basic_blocks f) = 1);
  let bb = Function.entry_block f in
  let num_args = Array.length (Function.params f) in
  assert (Array.length operands = num_args);
  let v_map = Hashtbl.create num_args in
  Array.iteri (fun i arg ->
    let operand = operands.(i) in
    assert (Llvm_analysis.cast_is_valid Bitcast operand (Type.type_of arg));
    let operand = Llvm.constrain_cast Bitcast operand (Type.type_of arg) "" intrinsic_builder.builder in
    Hashtbl.add v_map arg operand
  ) (Function.params f);
  let ret_val = ref None in
  List.iter (fun i ->
    match i with
    | Ret ret -> ret_val := Some (Return.value ret)
    | _ ->
      let new_i = Instruction.clone i in
      Instruction.insert new_i "" intrinsic_builder.builder;
      Hashtbl.add v_map i new_i;
      Instruction.remap_instruction new_i v_map [NoModuleLevelChanges; IgnoreMissingLocals];
      match new_i with
      | Call ci ->
        let callee = Call.called_function ci in
        assert (Function.is_intrinsic callee);
        let m = Instruction.parent new_i |> Basic_block.parent |> Function.parent in
        let intrinsic_decl = Module.declare_function m (Function.name callee) (Function.type_of callee) in
        Call.set_called_function ci intrinsic_decl
      | _ -> ()
  ) (Basic_block.instructions bb);
  match !ret_val with
  | Some ret_val ->
    let output = Hashtbl.find v_map ret_val in
    assert (Option.is_some output);
    Option.get output
  | None -> failwith "Wrapper not returning explicitly"

File: go/vectorize/intrinsic_builder.go
package hyper

import (
  "fmt"
  "github.com/llir/llvm/ir"
  "github.com/llir/llvm/ir/constant"
  "github.com/llir/llvm/ir/types"
  "github.com/llir/llvm/ir/value"
)

type IntrinsicBuilder struct {
  InstWrappers *ir.Module
  Builder      *ir.Builder
}

func (ib *IntrinsicBuilder) Create(name string, operands []value.Value, imm8 uint8) (value.Value, error) {
  wrapperName := fmt.Sprintf("intrinsic_wrapper_%s_%d", name, imm8)
  f := ib.InstWrappers.Func(wrapperName)
  if f == nil {
    return nil, fmt.Errorf("intrinsic wrapper undefined")
  }
  if len(f.Blocks) != 1 {
    return nil, fmt.Errorf("intrinsic wrapper should have a single basic block")
  }
  bb := f.Blocks[0]
  if len(operands) != len(f.Params) {
    return nil, fmt.Errorf("incorrect number of operands")
  }

  vMap := make(map[value.Value]value.Value)
  for i, arg := range f.Params {
    if !constant.IsValidCast(constant.BitCast, operands[i], arg.Type()) {
      return nil, fmt.Errorf("invalid input type")
    }
    operand := ib.Builder.CreateBitCast(operands[i], arg.Type(), "")
    vMap[arg] = operand
  }

  var retVal value.Value
  for _, inst := range bb.Insts {
    if ret, ok := inst.(*ir.InstRet); ok {
      retVal = ret.X
      break
    }
    newI := inst.Clone()
    ib.Builder.Insert(newI)
    vMap[inst] = newI
    remapInstruction(newI, vMap)
    if call, ok := newI.(*ir.InstCall); ok {
      callee := call.Callee
      if f, ok := callee.(*ir.Func); ok && f.IsIntrinsic() {
        m := ib.Builder.GetInsertBlock().Parent.Module
        intrinsicDecl := m.NewFunc(f.Name(), f.Sig)
        call.Callee = intrinsicDecl
      }
    }
  }

  if retVal == nil {
    return nil, fmt.Errorf("wrapper not returning explicitly")
  }
  output, ok := vMap[retVal]
  if !ok {
    return nil, fmt.Errorf("return value not found in value map")
  }
  return output, nil
}

func remapInstruction(inst ir.Instruction, vMap map[value.Value]value.Value) {}